src/client/Client.js

```diff
@@ -252,24 +207,65 @@ class Client extends BaseClient {
    * @example
    * client.login('my token');
    */
-  login(token = this.token) {
-    return new Promise((resolve, reject) => {
-      if (!token || typeof token !== 'string') throw new Error('TOKEN_INVALID');
-      token = token.replace(/^Bot\s*/i, '');
-      this.manager.connectToWebSocket(token, resolve, reject);
-    }).catch(e => {
-      this.destroy();
-      return Promise.reject(e);
+  async login(token = this.token) {
+    if (!token || typeof token !== 'string') throw new Error('TOKEN_INVALID');
+    this.token = token = token.replace(/^(Bot|Bearer)\s*/i, '');
+    this.emit(Events.DEBUG, `Authenticating using token ${token}`);
+    let endpoint = this.api.gateway;
+    if (this.options.shardCount === 'auto') endpoint = endpoint.bot;
+    const res = await endpoint.get();
+    if (this.options.presence) {
+      this.options.ws.presence = await this.presence._parse(this.options.presence);
+    }
+    if (res.session_start_limit && res.session_start_limit.remaining === 0) {
+      const { session_start_limit: { reset_after } } = res;
+      this.emit(Events.DEBUG, `Exceeded identify threshold, setting a timeout for ${reset_after} ms`);
+      await delayFor(reset_after);
+    }
+    const gateway = `${res.url}/`;
+    if (this.options.shardCount === 'auto') {
+      this.emit(Events.DEBUG, `Using recommended shard count ${res.shards}`);
+      this.options.shardCount = res.shards;
+      this.options.totalShardCount = res.shards;
+    }
+    this.emit(Events.DEBUG, `Using gateway ${gateway}`);
+    this.ws.connect(gateway);
+    await new Promise((resolve, reject) => {
+      const onready = () => {
+        clearTimeout(timeout);
+        this.removeListener(Events.DISCONNECT, ondisconnect);
+        resolve();
+      };
+      const ondisconnect = event => {
+        clearTimeout(timeout);
+        this.removeListener(Events.READY, onready);
+        this.removeListener(Events.DISCONNECT, ondisconnect);
```

Comment:
Doing this is redundant, the `ondisconnect` listener was only added once, and is no longer attached anyway.